cpp/src/arrow/compute/kernels/scalar_boolean.cc

```diff
@@ -129,13 +222,51 @@ struct Or {
   }
 };
 
-struct KleeneOr {
+struct KleeneOr : Commutative<KleeneOr> {
+  using Commutative<KleeneOr>::Call;
+
+  static void Call(KernelContext* ctx, const Scalar& left, const Scalar& right,
+                   Scalar* out) {
+    bool left_true = left.is_valid && checked_cast<const BooleanScalar&>(left).value;
+    bool left_false = left.is_valid && !checked_cast<const BooleanScalar&>(left).value;
+
+    bool right_true = right.is_valid && checked_cast<const BooleanScalar&>(right).value;
+    bool right_false = right.is_valid && !checked_cast<const BooleanScalar&>(right).value;
+
+    checked_cast<BooleanScalar*>(out)->value = left_true || right_true;
+    out->is_valid = left_true || right_true || (left_false && right_false);
+  }
+
+  static void Call(KernelContext* ctx, const ArrayData& left, const Scalar& right,
+                   ArrayData* out) {
+    bool right_true = right.is_valid && checked_cast<const BooleanScalar&>(right).value;
+    bool right_false = right.is_valid && !checked_cast<const BooleanScalar&>(right).value;
+
+    if (right_true) {
+      return GetBitmap(*out, 0).SetBitsTo(true),
+             GetBitmap(*out, 1).SetBitsTo(true);  // all true case
+    }
+
+    if (right_false) {
+      return GetBitmap(*out, 0).CopyFrom(GetBitmap(left, 0)),
+             GetBitmap(*out, 1).CopyFrom(GetBitmap(left, 1));
+    }
+
+    // scalar was null: out[i] is valid iff left[i] was true
+    ::arrow::internal::BitmapAnd(left.buffers[0]->data(), left.offset,
+                                 left.buffers[1]->data(), left.offset, left.length,
+                                 out->offset, out->buffers[0]->mutable_data());
+    ::arrow::internal::CopyBitmap(left.buffers[1]->data(), left.offset, left.length,
+                                  out->buffers[1]->mutable_data(), out->offset);
+  }
+
   static void Call(KernelContext* ctx, const ArrayData& left, const ArrayData& right,
                    ArrayData* out) {
     if (left.GetNullCount() == 0 && right.GetNullCount() == 0) {
-      BitUtil::SetBitsTo(out->buffers[0]->mutable_data(), out->offset, out->length, true);
+      GetBitmap(*out, 0).SetBitsTo(true);
```

Comment:
Should you make the same change in `KleeneAnd::Call(KernelContext*, const ArrayData&, const ArrayData&, ArrayData*)`?